README.md

Pacer
=====

A flexible, fault-tolerant, Redis-based rate-limiter for Node.js.

[![NPM version][shield-npm]][info-npm]
[![Node.js version support][shield-node]][info-node]
[![Build status][shield-build]][info-build]
[![Dependencies][shield-dependencies]][info-dependencies]
[![MIT licensed][shield-license]][info-license]

```js
var createPacer = require('pacer');

var pacer = createPacer({
    limit: 5,  // Allow 5 requests...
    reset: 10  // ...every 10 seconds
});

pacer.consume('consumer-identifier', function (consumer) {
    // consumer == {
    //     id: 'consumer-identifier',
    //     limit: 5,
    //     remaining: 4,
    //     reset: 10,
    //     allowed: true
    // }
});
```

Table Of Contents
-----------------

- [Install](#install)
- [Getting Started](#getting-started)
- [Usage](#usage)
- [Options](#options)
- [Examples](#examples)
- [Contributing](#contributing)
- [License](#license)


Install
-------

Install Pacer with [npm][npm]:

```sh
npm install pacer
```


Getting Started
---------------

Require in and Pacer:

```js
var createPacer = require('pacer');
```

Create a Pacer instance, passing in some [options](#options):

```js
var pacer = createPacer({
    limit: 5,  // Allow 5 requests...
    reset: 10  // ...every 10 seconds
});
```

Consume a token for the `'foo'` consumer. The consumer identifier can be any string you like, for example the request IP, or a username.

```js
pacer.consume('foo', function (consumer) {
    // consumer == {
    //     id: 'foo',
    //     limit: 5,
    //     remaining: 4,
    //     reset: 10,
    //     allowed: true
    // }
});
```

Consumer can also be an object, allowing you to specify a custom limit and reset time for them. This could be used to provide different tiers of rate limiting depending on the user that is accessing your application.

```js
pacer.consume({
    id: 'foo',
    limit: 10,  // Allow 10 requests...
    reset: 5    // ...every 5 seconds
}, function (consumer) {
    // consumer == {
    //     id: 'consumer-identifier',
    //     limit: 10,
    //     remaining: 9,
    //     reset: 5,
    //     allowed: true
    // }
});
```

If you don't wish to consume a token but want to query how many tokens a consumer has remaining, you can use the `query` method. This works with string or object consumers.

```js
pacer.query('foo', function (consumer) {
    // consumer == {
    //     id: 'foo',
    //     limit: 5,
    //     remaining: 5,
    //     reset: 10,
    //     allowed: true
    // }
});
```


Usage
-----

Create a pacer with the passed in [`options`](#options) object:

```js
var pacer = createPacer({ /* ... */ });
```

### Consuming

#### `pacer.consume(consumerId, callback)`

Consume a token for a specified consumer.

```js
consumerId = String
callback = function (consumer) {
    consumer = {
        id: String,
        limit: Number,
        remaining: Number,
        reset: Number,
        allowed: Boolean
    }
}
```

#### `pacer.consume(consumer, callback)`

Consume a token for a specified consumer using a custom limit and reset time.

```js
consumer = {
    id: String,
    limit: Number,
    reset: Number
}
callback = function (consumer) {
    consumer = {
        id: String,
        limit: Number,
        remaining: Number,
        reset: Number,
        allowed: Boolean
    }
}
```

### Querying

#### `pacer.query(consumerId, callback)`

Query the tokens for a specified consumer.

```js
consumerId = String
callback = function (consumer) {
    consumer = {
        id: String,
        limit: Number,
        remaining: Number,
        reset: Number,
        allowed: Boolean
    }
}
```

#### `pacer.query(consumer, callback)`

Query the tokens for a specified consumer using a custom limit and reset time.

```js
consumer = {
    id: String,
    limit: Number,
    reset: Number
}
callback = function (consumer) {
    consumer = {
        id: String,
        limit: Number,
        remaining: Number,
        reset: Number,
        allowed: Boolean
    }
}
```


Options
-------

#### `allowOnError` (boolean)

Whether to allow access for all consumers if the Redis server errors or is down. Defaults to `true`.

#### `limit` (number)

The maximum number of tokens a consumer can use. Defaults to `100`.

#### `redisHost` (string)

The host Redis is running on. Defaults to `'localhost'`.

#### `redisIndex` (number)

The Redis database index to use. Defaults to `0`.

#### `redisPort` (number)

The port Redis is running on. Defaults to `6379`.

#### `reset` (number)

The amount of time (in seconds) before tokens for a consumer are reset to their maximum. Defaults to `3600`.


Examples
--------

Pacer comes with a few examples which demonstrate how you can integrate it with your applications:

#### Basic Example

A simple command-line interface which consumes a rate-limiting token for the given consumer every time it's called.

```
node example/basic myconsumerid
```

#### Query Example

A simple command-line interface which queries the rate-limiting tokens for the given consumer every time it's called, without consuming one.

```
node example/query myconsumerid
```

#### Basic Connect Example

A connect application which consumes a rate-limiting token for the client IP whenever a request it made.

```
node example/connect-basic
```

Then open http://localhost:3000/ in your browser.

#### Connect With Varying Rates Example

A connect application which consumes a rate-limiting token for the client IP whenever a request it made. Google Chrome users get a higher rate limit and a faster reset time, to demonstrate varying rates for different consumer types.

```
node example/connect-varying-rates
```

Then open http://localhost:3000/ in your browser.


Contributing
------------

To contribute to Pacer, clone this repo locally and commit your code on a separate branch.

Please write unit tests for your code, and check that everything works by running the following before opening a pull-request:

```sh
make ci
```


License
-------

Pacer is licensed under the [MIT][info-license] license.  
Copyright &copy; 2015, Rowan Manning



[npm]: https://npmjs.org/

[info-dependencies]: https://gemnasium.com/rowanmanning/pacer
[info-license]: LICENSE
[info-node]: package.json
[info-npm]: https://www.npmjs.com/package/pacer
[info-build]: https://travis-ci.org/rowanmanning/pacer
[shield-dependencies]: https://img.shields.io/gemnasium/rowanmanning/pacer.svg
[shield-license]: https://img.shields.io/badge/license-MIT-blue.svg
[shield-node]: https://img.shields.io/badge/node.js%20support-0.10–4-brightgreen.svg
[shield-npm]: https://img.shields.io/npm/v/pacer.svg
[shield-build]: https://img.shields.io/travis/rowanmanning/pacer/master.svg


HISTORY.md

# History

## 0.1.2 pre-release (2015-09-17)

  * Add Node.js 4.x support
  * Add test coverage reporting
  * Update dependencies

## 0.1.1 pre-release (2015-06-21)

  * Update dependencies
  * Use JSCS to check for code-style issues

## 0.1.0 pre-release (2015-05-03)

  * Initial release


test/unit/setup.js
/* jshint maxstatements: false, maxlen: false */
/* global afterEach, beforeEach */
'use strict';

var assert = require('proclaim');
var mockery = require('mockery');
var sinon = require('sinon');

sinon.assert.expose(assert, {
    includeFail: false,
    prefix: ''
});

beforeEach(function () {
    mockery.enable({
        useCleanCache: true,
        warnOnUnregistered: false,
        warnOnReplace: false
    });
});

afterEach(function () {
    mockery.deregisterAll();
    mockery.disable();
});


test/unit/lib/pacer.js
// jshint maxstatements: false
// jscs:disable disallowMultipleVarDecl, maximumLineLength
'use strict';

var assert = require('proclaim');
var mockery = require('mockery');

describe('lib/pacer', function () {
    var createPacer, redis, underscore;

    beforeEach(function () {

        redis = require('../mock/redis');
        mockery.registerMock('redis', redis);

        underscore = require('../mock/underscore');
        mockery.registerMock('underscore', underscore);

        createPacer = require('../../../lib/pacer');

    });

    it('should be a function', function () {
        assert.isFunction(createPacer);
    });

    it('should have a `defaults` property', function () {
        assert.isObject(createPacer.defaults);
    });

    describe('.defaults', function () {
        var defaults;

        beforeEach(function () {
            defaults = createPacer.defaults;
        });

        it('should have an `allowOnError` property', function () {
            assert.isTrue(defaults.allowOnError);
        });

        it('should have a `redisHost` property', function () {
            assert.strictEqual(defaults.redisHost, 'localhost');
        });

        it('should have a `redisPort` property', function () {
            assert.strictEqual(defaults.redisPort, 6379);
        });

        it('should have a `redisIndex` property', function () {
            assert.strictEqual(defaults.redisIndex, 0);
        });

        it('should have a `limit` property', function () {
            assert.strictEqual(defaults.limit, 100);
        });

        it('should have a `reset` property', function () {
            assert.strictEqual(defaults.reset, 3600);
        });

    });

    describe('createPacer()', function () {
        var options, pacer, redisClient;

        beforeEach(function () {
            options = {
                allowOnError: true,
                redisHost: 'localhost',
                redisPort: 6379,
                redisIndex: 0,
                limit: 100,
                reset: 3600
            };
            pacer = createPacer(options);
            redisClient = redis.createClient.firstCall.returnValue;
        });

        it('should default the options', function () {
            assert.isTrue(underscore.defaults.calledOnce);
            assert.deepEqual(underscore.defaults.firstCall.args[0], {});
            assert.strictEqual(underscore.defaults.firstCall.args[1], options);
            assert.strictEqual(underscore.defaults.firstCall.args[2], createPacer.defaults);
        });

        it('should create a Redis client', function () {
            assert.isTrue(redis.createClient.withArgs(options.redisPort, options.redisHost).calledOnce);
        });

        it('should select a Redis database', function () {
            assert.isTrue(redisClient.select.withArgs(options.redisIndex).calledOnce);
        });

        it('should return an object', function () {
            assert.isObject(pacer);
        });

        describe('returned object', function () {

            it('should have a `consume` method', function () {
                assert.isFunction(pacer.consume);
            });

            describe('.consume() with a string consumer', function () {
                var consumer, multi;

                beforeEach(function (done) {
                    consumer = 'foo';
                    pacer.consume(consumer, function () {
                        multi = redisClient.multi.firstCall.returnValue;
                        done();
                    });
                });

                it('should create a Redis multi command', function () {
                    assert.isTrue(redisClient.multi.calledOnce);
                });

                it('should SET the consumer key in Redis', function () {
                    assert.isTrue(multi.set.calledOnce);
                    assert.deepEqual(multi.set.firstCall.args[0], [
                        consumer,
                        options.limit,
                        'NX',
                        'EX',
                        options.reset
                    ]);
                });

                it('should DECR the consumer key in Redis', function () {
                    assert.isTrue(multi.decr.withArgs(consumer).calledOnce);
                });

                it('should not GET the consumer key in Redis', function () {
                    assert.isFalse(multi.get.called);
                });

                it('should get the TTL of the consumer key in Redis', function () {
                    assert.isTrue(multi.ttl.withArgs(consumer).calledOnce);
                });

                it('should execute the multi command with a callback', function () {
                    assert.isTrue(multi.exec.calledOnce);
                    assert.isFunction(multi.exec.firstCall.args[0]);
                });

                it('should call Redis multi functions in order', function () {
                    assert.callOrder(
                        redisClient.multi,
                        multi.set,
                        multi.decr,
                        multi.ttl,
                        multi.exec
                    );
                });

            });

            describe('.consume() with an object consumer', function () {
                var consumer, multi;

                beforeEach(function (done) {
                    consumer = {
                        id: 'foo',
                        limit: 1234,
                        reset: 5678
                    };
                    pacer.consume(consumer, function () {
                        multi = redisClient.multi.firstCall.returnValue;
                        done();
                    });
                });

                it('should create a Redis multi command', function () {
                    assert.isTrue(redisClient.multi.calledOnce);
                });

                it('should SET the consumer key in Redis', function () {
                    assert.isTrue(multi.set.calledOnce);
                    assert.deepEqual(multi.set.firstCall.args[0], [
                        consumer.id,
                        consumer.limit,
                        'NX',
                        'EX',
                        consumer.reset
                    ]);
                });

                it('should DECR the consumer key in Redis', function () {
                    assert.isTrue(multi.decr.withArgs(consumer.id).calledOnce);
                });

                it('should not GET the consumer key in Redis', function () {
                    assert.isFalse(multi.get.called);
                });

                it('should get the TTL of the consumer key in Redis', function () {
                    assert.isTrue(multi.ttl.withArgs(consumer.id).calledOnce);
                });

                it('should execute the multi command with a callback', function () {
                    assert.isTrue(multi.exec.calledOnce);
                    assert.isFunction(multi.exec.firstCall.args[0]);
                });

                it('should call Redis multi functions in order', function () {
                    assert.callOrder(
                        redisClient.multi,
                        multi.set,
                        multi.decr,
                        multi.ttl,
                        multi.exec
                    );
                });

            });

            describe('.consume() with an object consumer that has no defined limit', function () {
                var consumer, multi;

                beforeEach(function (done) {
                    consumer = {
                        id: 'foo',
                        reset: 5678
                    };
                    pacer.consume(consumer, function () {
                        multi = redisClient.multi.firstCall.returnValue;
                        done();
                    });
                });

                it('should SET the consumer key in Redis with the default limit', function () {
                    assert.isTrue(multi.set.calledOnce);
                    assert.strictEqual(multi.set.firstCall.args[0][1], options.limit);
                });

            });

            describe('.consume() with an object consumer that has no defined reset', function () {
                var consumer, multi;

                beforeEach(function (done) {
                    consumer = {
                        id: 'foo',
                        limit: 1234
                    };
                    pacer.consume(consumer, function () {
                        multi = redisClient.multi.firstCall.returnValue;
                        done();
                    });
                });

                it('should SET the consumer key in Redis with the default reset', function () {
                    assert.isTrue(multi.set.calledOnce);
                    assert.strictEqual(multi.set.firstCall.args[0][4], options.reset);
                });

            });

            describe('.consume() result handler', function () {
                var exec;

                beforeEach(function () {
                    exec = redisClient.multi.defaultBehavior.returnValue.exec;
                });

                describe('when there are tokens remaining', function () {
                    var consumer, resultHandler;

                    beforeEach(function (done) {
                        exec.yields(null, [
                            null,
                            '12', // remaining
                            '345' // reset
                        ]);
                        pacer.consume('foo', function (result) {
                            consumer = result;
                            resultHandler = exec.firstCall.args[0];
                            done();
                        });
                    });

                    it('should callback with the expected consumer object', function () {
                        assert.strictEqual(consumer.id, 'foo');
                        assert.strictEqual(consumer.limit, 100);
                        assert.strictEqual(consumer.reset, 345);
                        assert.strictEqual(consumer.remaining, 12);
                        assert.isNull(consumer.error);
                        assert.isTrue(consumer.allowed);
                    });

                });

                describe('when no more tokens are remaining', function () {
                    var consumer, resultHandler;

                    beforeEach(function (done) {
                        exec.yields(null, [
                            null,
                            '0' // remaining
                        ]);
                        pacer.consume('foo', function (result) {
                            consumer = result;
                            resultHandler = exec.firstCall.args[0];
                            done();
                        });
                    });

                    it('should callback with the expected consumer object', function () {
                        assert.strictEqual(consumer.remaining, 0);
                        assert.isNull(consumer.error);
                        assert.isFalse(consumer.allowed);
                    });

                });

                describe('when remaining tokens are negative', function () {
                    var consumer, resultHandler;

                    beforeEach(function (done) {
                        exec.yields(null, [
                            null,
                            '-1' // remaining
                        ]);
                        pacer.consume('foo', function (result) {
                            consumer = result;
                            resultHandler = exec.firstCall.args[0];
                            done();
                        });
                    });

                    it('should callback with the expected consumer object', function () {
                        assert.strictEqual(consumer.remaining, 0);
                        assert.isNull(consumer.error);
                        assert.isFalse(consumer.allowed);
                    });

                });

                describe('when the Redis query errors and `options.allowOnError` is `true`', function () {
                    var error, consumer, resultHandler;

                    beforeEach(function (done) {
                        error = new Error('...');
                        exec.yields(error, null);
                        pacer.consume('foo', function (result) {
                            consumer = result;
                            resultHandler = exec.firstCall.args[0];
                            done();
                        });
                    });

                    it('should callback with the expected consumer object', function () {
                        assert.strictEqual(consumer.id, 'foo');
                        assert.strictEqual(consumer.limit, 100);
                        assert.strictEqual(consumer.reset, 0);
                        assert.strictEqual(consumer.remaining, 100);
                        assert.strictEqual(consumer.error, error);
                        assert.isTrue(consumer.allowed);
                    });

                });

                describe('when the Redis query errors and `options.allowOnError` is `false`', function () {
                    var error, consumer, resultHandler;

                    beforeEach(function (done) {
                        error = new Error('...');
                        exec.yields(error, null);
                        options.allowOnError = false;
                        pacer = createPacer(options);
                        pacer.consume('foo', function (result) {
                            consumer = result;
                            resultHandler = exec.firstCall.args[0];
                            done();
                        });
                    });

                    it('should callback with the expected consumer object', function () {
                        assert.strictEqual(consumer.id, 'foo');
                        assert.strictEqual(consumer.limit, 100);
                        assert.strictEqual(consumer.reset, 0);
                        assert.strictEqual(consumer.remaining, 0);
                        assert.strictEqual(consumer.error, error);
                        assert.isFalse(consumer.allowed);
                    });

                });

            });

            it('should have a `query` method', function () {
                assert.isFunction(pacer.query);
            });

            describe('.query() with a string consumer', function () {
                var consumer, multi;

                beforeEach(function (done) {
                    consumer = 'foo';
                    pacer.query(consumer, function () {
                        multi = redisClient.multi.firstCall.returnValue;
                        done();
                    });
                });

                it('should create a Redis multi command', function () {
                    assert.isTrue(redisClient.multi.calledOnce);
                });

                it('should SET the consumer key in Redis', function () {
                    assert.isTrue(multi.set.calledOnce);
                    assert.deepEqual(multi.set.firstCall.args[0], [
                        consumer,
                        options.limit,
                        'NX',
                        'EX',
                        options.reset
                    ]);
                });

                it('should GET the consumer key in Redis', function () {
                    assert.isTrue(multi.get.withArgs(consumer).calledOnce);
                });

                it('should not DECR the consumer key in Redis', function () {
                    assert.isFalse(multi.decr.called);
                });

                it('should get the TTL of the consumer key in Redis', function () {
                    assert.isTrue(multi.ttl.withArgs(consumer).calledOnce);
                });

                it('should execute the multi command with a callback', function () {
                    assert.isTrue(multi.exec.calledOnce);
                    assert.isFunction(multi.exec.firstCall.args[0]);
                });

                it('should call Redis multi functions in order', function () {
                    assert.callOrder(
                        redisClient.multi,
                        multi.set,
                        multi.get,
                        multi.ttl,
                        multi.exec
                    );
                });

            });

            describe('.query() with an object consumer', function () {
                var consumer, multi;

                beforeEach(function (done) {
                    consumer = {
                        id: 'foo',
                        limit: 1234,
                        reset: 5678
                    };
                    pacer.query(consumer, function () {
                        multi = redisClient.multi.firstCall.returnValue;
                        done();
                    });
                });

                it('should create a Redis multi command', function () {
                    assert.isTrue(redisClient.multi.calledOnce);
                });

                it('should SET the consumer key in Redis', function () {
                    assert.isTrue(multi.set.calledOnce);
                    assert.deepEqual(multi.set.firstCall.args[0], [
                        consumer.id,
                        consumer.limit,
                        'NX',
                        'EX',
                        consumer.reset
                    ]);
                });

                it('should GET the consumer key in Redis', function () {
                    assert.isTrue(multi.get.withArgs(consumer.id).calledOnce);
                });

                it('should not DECR the consumer key in Redis', function () {
                    assert.isFalse(multi.decr.called);
                });

                it('should get the TTL of the consumer key in Redis', function () {
                    assert.isTrue(multi.ttl.withArgs(consumer.id).calledOnce);
                });

                it('should execute the multi command with a callback', function () {
                    assert.isTrue(multi.exec.calledOnce);
                    assert.isFunction(multi.exec.firstCall.args[0]);
                });

                it('should call Redis multi functions in order', function () {
                    assert.callOrder(
                        redisClient.multi,
                        multi.set,
                        multi.get,
                        multi.ttl,
                        multi.exec
                    );
                });

            });

            describe('.query() result handler', function () {
                var exec;

                beforeEach(function () {
                    exec = redisClient.multi.defaultBehavior.returnValue.exec;
                });

                describe('when there are tokens remaining', function () {
                    var consumer, resultHandler;

                    beforeEach(function (done) {
                        exec.yields(null, [
                            null,
                            '12', // remaining
                            '345' // reset
                        ]);
                        pacer.query('foo', function (result) {
                            consumer = result;
                            resultHandler = exec.firstCall.args[0];
                            done();
                        });
                    });

                    it('should callback with the expected consumer object', function () {
                        assert.strictEqual(consumer.id, 'foo');
                        assert.strictEqual(consumer.limit, 100);
                        assert.strictEqual(consumer.reset, 345);
                        assert.strictEqual(consumer.remaining, 12);
                        assert.isNull(consumer.error);
                        assert.isTrue(consumer.allowed);
                    });

                });

                describe('when no more tokens are remaining', function () {
                    var consumer, resultHandler;

                    beforeEach(function (done) {
                        exec.yields(null, [
                            null,
                            '0' // remaining
                        ]);
                        pacer.query('foo', function (result) {
                            consumer = result;
                            resultHandler = exec.firstCall.args[0];
                            done();
                        });
                    });

                    it('should callback with the expected consumer object', function () {
                        assert.strictEqual(consumer.remaining, 0);
                        assert.isNull(consumer.error);
                        assert.isFalse(consumer.allowed);
                    });

                });

                describe('when remaining tokens are negative', function () {
                    var consumer, resultHandler;

                    beforeEach(function (done) {
                        exec.yields(null, [
                            null,
                            '-1' // remaining
                        ]);
                        pacer.query('foo', function (result) {
                            consumer = result;
                            resultHandler = exec.firstCall.args[0];
                            done();
                        });
                    });

                    it('should callback with the expected consumer object', function () {
                        assert.strictEqual(consumer.remaining, 0);
                        assert.isNull(consumer.error);
                        assert.isFalse(consumer.allowed);
                    });

                });

                describe('when the Redis query errors and `options.allowOnError` is `true`', function () {
                    var error, consumer, resultHandler;

                    beforeEach(function (done) {
                        error = new Error('...');
                        exec.yields(error, null);
                        pacer.query('foo', function (result) {
                            consumer = result;
                            resultHandler = exec.firstCall.args[0];
                            done();
                        });
                    });

                    it('should callback with the expected consumer object', function () {
                        assert.strictEqual(consumer.id, 'foo');
                        assert.strictEqual(consumer.limit, 100);
                        assert.strictEqual(consumer.reset, 0);
                        assert.strictEqual(consumer.remaining, 100);
                        assert.strictEqual(consumer.error, error);
                        assert.isTrue(consumer.allowed);
                    });

                });

                describe('when the Redis query errors and `options.allowOnError` is `false`', function () {
                    var error, consumer, resultHandler;

                    beforeEach(function (done) {
                        error = new Error('...');
                        exec.yields(error, null);
                        options.allowOnError = false;
                        pacer = createPacer(options);
                        pacer.query('foo', function (result) {
                            consumer = result;
                            resultHandler = exec.firstCall.args[0];
                            done();
                        });
                    });

                    it('should callback with the expected consumer object', function () {
                        assert.strictEqual(consumer.id, 'foo');
                        assert.strictEqual(consumer.limit, 100);
                        assert.strictEqual(consumer.reset, 0);
                        assert.strictEqual(consumer.remaining, 0);
                        assert.strictEqual(consumer.error, error);
                        assert.isFalse(consumer.allowed);
                    });

                });

            });

        });

    });

});


test/unit/mock/redis.js
'use strict';

var sinon = require('sinon');

module.exports = {
    createClient: sinon.stub().returns({
        select: sinon.spy(),
        multi: sinon.stub().returns({
            decr: sinon.spy(),
            exec: sinon.stub().yields(),
            get: sinon.spy(),
            set: sinon.spy(),
            ttl: sinon.spy()
        })
    })
};


lib/pacer.js
'use strict';

var _ = require('underscore');
var redis = require('redis');

module.exports = createPacer;
module.exports.defaults = {
    allowOnError: true,
    redisHost: 'localhost',
    redisPort: 6379,
    redisIndex: 0,
    limit: 100,
    reset: 3600
};

function createPacer (options) {
    options = defaultOptions(options);
    var pacer = {
        database: createRedisClient(
            options.redisHost,
            options.redisPort,
            options.redisIndex
        ),
        limit: options.limit,
        reset: options.reset,
        allowOnError: Boolean(options.allowOnError)
    };
    return {
        consume: consumeRateLimit.bind(null, pacer),
        query: queryRateLimit.bind(null, pacer)
    };
}

function createRedisClient (host, port, database) {
    var client = redis.createClient(port, host);
    client.select(database);
    return client;
}

function consumeRateLimit (pacer, consumer, done) {
    performQuery(pacer, consumer, true, done);
}

function queryRateLimit (pacer, consumer, done) {
    performQuery(pacer, consumer, false, done);
}

function performQuery (pacer, consumer, shouldConsume, done) {
    consumer = resolveConsumer(consumer, pacer);
    var multi = pacer.database.multi();
    multi.set([consumer.id, consumer.limit, 'NX', 'EX', consumer.reset]);
    if (shouldConsume) {
        multi.decr(consumer.id);
    }
    else {
        multi.get(consumer.id);
    }
    multi.ttl(consumer.id);
    multi.exec(function (error, results) {
        if (error || !results) {
            consumer.remaining = (pacer.allowOnError ? consumer.limit : 0);
            consumer.reset = 0;
        }
        else {
            consumer.remaining = Math.max(0, parseInt(results[1], 10));
            consumer.reset = parseInt(results[2], 10);
        }
        consumer.error = error;
        consumer.allowed = (consumer.remaining > 0);
        done(consumer);
    });
}

function resolveConsumer (consumer, pacer) {
    return {
        id: resolveConsumerId(consumer),
        limit: parseInt(resolveConsumerLimit(consumer, pacer), 10),
        reset: parseInt(resolveConsumerReset(consumer, pacer), 10)
    };
}

function resolveConsumerId (consumer) {
    if (isObject(consumer)) {
        return consumer.id;
    }
    return String(consumer);
}

function resolveConsumerLimit (consumer, pacer) {
    if (isObject(consumer)) {
        return consumer.limit || pacer.limit;
    }
    return pacer.limit;
}

function resolveConsumerReset (consumer, pacer) {
    if (isObject(consumer)) {
        return consumer.reset || pacer.reset;
    }
    return pacer.reset;
}

function defaultOptions (options) {
    return _.defaults({}, options, createPacer.defaults);
}

function isObject (value) {
    return (value !== null && !Array.isArray(value) && typeof value === 'object');
}
